File: 1107-minimum-swaps-to-group-all-1s-together/1107-minimum-swaps-to-group-all-1s-together.java
class Solution {
    public int minSwaps(int[] data) {

        int ones = Arrays.stream(data).sum();
        int cnt_one = 0, max_one =0;

        int left= 0, right = 0;
        // loop through each element of array starting from 0
        while (right < data.length)
        {
            // Keep on adding current value to total 1 encountered so far
            cnt_one += data[right++];

            // To window slide in sub-array with size of max 1s in array
            if(right - left > ones){
                cnt_one -= data[left++]; 
            }

            max_one = Math.max(max_one, cnt_one);
        }
        return ones - max_one; 
        
    }
}

File: 1107-minimum-swaps-to-group-all-1s-together/1107-minimum-swaps-to-group-all-1s-together.py
class Solution:
    def minSwaps(self, data: List[int]) -> int:

        # Calculate number of one in list
        count = sum(data)
        count_one = swaps = 0

        if count == 1:
            return swaps

        left = right = 0

        # Check for sub array to bring all 1s togther
        while right < len(data):
            # Increase the sum of ones
            count_one += data[right]
            # increment the pointer
            right += 1

            if right - left > count:
                count_one -= data[left]
                left += 1

            swaps = max(swaps, count_one)


        return count - swaps